### PR TITLE
Stricter SPIFFE ID and trust domain name validation

### PR DESCRIPTION
**Pull request by azdagron**
This change updates the existing validation checks to conform to the updated character set requirements for SPIFFE ID trust domain and path components. Although go-spiffe v2 is getting a similar treatment, we cannot not rely on updates to that library in order to preserve the safety valve that we are keeping in place for 1.0.0. The safety valve will be removed in 1.1, and the idutil parsing and validation functions removed/rewritten in favor of the updated go-spiffe v2 library.

This change also updates the bundle create/update/set RPCs to validate the incoming trust domain against the stricter rules (subject to the safety valve).

### CODE REVIEW

File: pkg/common/idutil/safety.go [hunk 1]
```diff
@@ -176,18 +144,128 @@ func IDFromProto(id *types.SPIFFEID) (spiffeid.ID, error) {
...
+	return nil
+}
+
+func isValidTrustDomainChar(c uint8) bool {
+	switch {
+	case c >= 'a' && c <= 'z':
+		return true
+	case c >= '0' && c <= '9':
+		return true
+	case c == '.', c == '-':
```
Comment (evan2645) [inline pkg/common/idutil/safety.go:253]: We don't want underscore here too?

File: pkg/common/idutil/safety.go [hunk 2]
```diff
@@ -176,18 +144,130 @@ func IDFromProto(id *types.SPIFFEID) (spiffeid.ID, error) {
...
+	for i := 0; i < len(td); i++ {
+		if !isValidTrustDomainChar(td[i]) {
+			return errors.New("trust domain can only contain [a-z0-9.-]")
+		}
+	}
+	return nil
+}
+
+func validatePath(path string) error {
+	segment := 0
```
Comment (evan2645) [inline pkg/common/idutil/safety.go:219]: nit: `ii`? `index`? `starting_index`? `segment_index`? `last_segment_index`?

As is, the first thing I expect this to be is a count of segments, not the index at which to start a segment

File: pkg/common/idutil/safety.go [hunk 3]
```diff
@@ -176,18 +144,130 @@ func IDFromProto(id *types.SPIFFEID) (spiffeid.ID, error) {
...
+			case "/":
+				return errors.New("path cannot contain empty segments")
+			case "/.", "/..":
+				return errors.New("path cannot contain dot segments")
+			}
+			segment = i
+			continue
+		}
+		if !isValidPathSegmentChar(c) {
+			return errors.New("path can only contain [a-zA-Z0-9.-_]")
```
Comment (evan2645) [inline pkg/common/idutil/safety.go:234]:
```suggestion
			return errors.New("path can only contain letters, numbers, dots, dashes, and underscores")
```

While most SPIRE operators probably know regex, it's best to not expect them to. Same comment for `validateTrustDomain` and `TrustDomainFromString`

File: pkg/common/idutil/safety.go [hunk 4]
```diff
@@ -176,18 +144,130 @@ func IDFromProto(id *types.SPIFFEID) (spiffeid.ID, error) {
 // FormatPath formats a path string. The function ensures a leading slash is
 // present.
 func FormatPath(format string, args ...interface{}) string {
+	// TODO: when the safety valve is removed, this function should:
+	// 1. not ensure the leading slash.
+	// 2. validate that the produced path is correct
```
Comment (evan2645) [inline pkg/common/idutil/safety.go:149]: I am wondering about this ... shouldn't we do this with 1.0 by default, and only do `ensureLeadingSlash` if the safety valve is activated?

File: pkg/common/idutil/safety_test.go [hunk 5]
```diff
@@ -166,6 +174,10 @@ func testCommonCheckIDNormalization(assertGood func(string), assertBad func(stri
 	assertGood("spiffe://example.org/workload/%23")
 	assertGood("spiffe://example.org/workload/%00")
 	assertGood("spiffe://example.org/workload/" + url.PathEscape("%E4%B8%96%E7%95%8C"))
+	assertGood("spiffe://example.org/workload/%E4%B8%96%E7%95%8C")
+	assertGood("spiffe://世界/workload")
+	assertGood("spiffe://example.org/世界")
+	assertGood("spiffe://%E4%B8%96%E7%95%8C/workload")
```
Comment (evan2645) [inline pkg/common/idutil/safety_test.go:180]: I don't think we have any assertions that scheme rules are reverted with the safety valve?

**Review by evan2645 — COMMENTED** [review]
thank you so much for this @azdagron

it is looking pretty good. I will spend a little more time with the validation diff. Leaving some observations I made yesterday/today

File: pkg/common/idutil/safety.go (hunk 1 above)
Comment (azdagron) [reply pkg/common/idutil/safety.go:253]: Maybe? It wasn't clear. I couldn't find any reference in the meeting notes that indicated that we'd diverted from the open PR. I've fielded a question to SIG-SPEC.

File: pkg/common/idutil/safety_test.go (hunk 5 above)
Comment (azdagron) [reply pkg/common/idutil/safety_test.go:180]: :+1 I'll add some.

File: cmd/spire-server/cli/run/run.go [hunk 6]
```diff
@@ -330,6 +330,12 @@ func NewServerConfig(c *Config, logOptions []log.Option, allowUnknownConfig bool
 		return nil, err
 	}
 
+	// This is a terrible hack but is just a short-term band-aid.
+	if c.Server.AllowUnsafeIDs != nil {
+		sc.Log.Warn("The insecure allow_unsafe_ids configurable will be deprecated in a future release.")
```
[the remark this answers is not in the document]
Comment (azdagron) [reply cmd/spire-server/cli/run/run.go:336]: :+1:

File: pkg/common/idutil/safety.go (hunk 4 above)
Comment (azdagron) [reply pkg/common/idutil/safety.go:149]: We could. However, these two helpers are currently only used by the registrar, which doesn't have a way to enable the safety valve.

File: pkg/common/idutil/safety.go (hunk 2 above)
Comment (azdagron) [reply pkg/common/idutil/safety.go:219]: fixed

File: pkg/common/idutil/safety.go (hunk 3 above)
Comment (azdagron) [reply pkg/common/idutil/safety.go:234]: :+1:

File: cmd/spire-server/cli/run/run.go [hunk 7]
```diff
@@ -330,6 +331,12 @@ func NewServerConfig(c *Config, logOptions []log.Option, allowUnknownConfig bool
 		return nil, err
 	}
 
+	// This is a terrible hack but is just a short-term band-aid.
```
[the remark this answers is not in the document]
Comment (azdagron) [reply cmd/spire-server/cli/run/run.go:334]: :+1:

**Review by evan2645 — APPROVED** [review]

**Pull request merged**